### PR TITLE
Add flag to optionally average output attention weights across heads

### LINKED ISSUE #47583: multi_head_attention_forward: add option to return attention scores for all heads instead of averaging
## 🚀 Feature: Return the attention scores from all heads

Return the attention scores from all heads in the [`MultiHeadAttention`](https://github.com/pytorch/pytorch/blob/b5dd6e3e61c59e88ce6e8aee2fa5d08d6c56fd97/torch/nn/modules/activation.py#L831
) Module and [`multi_head_attention_forward`](https://github.com/pytorch/pytorch/blob/16c72a5a6b7085a9651e7bef7e98c80b2d5547e9/torch/nn/functional.py#L4026
). The current implementation **only** allows returning the averaged attention score over all heads 

In short: add option to change
https://github.com/pytorch/pytorch/blob/16c72a5a6b7085a9651e7bef7e98c80b2d5547e9/torch/nn/functional.py#L4309
into
`return attn_output, attn_output_weights`

## Motivation

The motivating idea behind Multi-Head attention is to perform the attention mechanism in parallel and allow the model to attend to different sequence elements with each head separately. The current torch code ([`MultiHeadAttention`](https://github.com/pytorch/pytorch/blob/b5dd6e3e61c59e88ce6e8aee2fa5d08d6c56fd97/torch/nn/modules/activation.py#L831
) Module and [`multi_head_attention_forward`](https://github.com/pytorch/pytorch/blob/16c72a5a6b7085a9651e7bef7e98c80b2d5547e9/torch/nn/functional.py#L4026
)) allows to return the `attn_output_weights` along with the `attn_output`. 

However, the `attn_output_weights` are averaged over all heads.  The attention scores for each separate head are not accessible to the outside.

Why should it be possible to get attention scores for each head?

1) Each head attends to different sequence elements. The information to which sequence elements each head attends to is lost when attention scores are averaged over all heads. The attention scores for each head are important to understand the features that are extracted in each layer (touches: feature importance, explainability).
2) This averaging behavior in `attn_output_weights` is inconsistent with `attn_output` where the embedding dimensions are grouped into `head_dim` blocks from the same attention head. 
https://github.com/pytorch/pytorch/blob/16c72a5a6b7085a9651e7bef7e98c80b2d5547e9/torch/nn/functional.py#L4124

## Pitch

I would propose 
1) modifying the `need_weights=True` option in [`multi_head_attention_forward`](https://github.com/pytorch/pytorch/blob/16c72a5a6b7085a9651e7bef7e98c80b2d5547e9/torch/nn/functional.py#L4026
) to a choice [`all`, `average`, `none`] to control the return behavior of `multi_head_attention_forward`. The option `need_weights=avg` would be equivalent to current `need_weights=True` and `need_weights=all` would return `attn_output_weights` without summation and average with `num_heads` [here](https://github.com/pytorch/pytorch/blob/16c72a5a6b7085a9651e7bef7e98c80b2d5547e9/torch/nn/functional.py#L4309). Alternatives like changing the current behavior or adding a second boolean `need_all_weights` flag are possible too.
2) carrying the `need_weights` option to the [`MultiHeadAttention`](https://github.com/pytorch/pytorch/blob/b5dd6e3e61c59e88ce6e8aee2fa5d08d6c56fd97/torch/nn/modules/activation.py#L831
) Module to make it easily accessible to the outside when using the `MultiHeadAttention` module. 

I am happy to write a pull request with the modifications in [`multi_head_attention_forward`](https://github.com/pytorch/pytorch/blob/16c72a5a6b7085a9651e7bef7e98c80b2d5547e9/torch/nn/functional.py#L4026
) and [`MultiHeadAttention`](https://github.com/pytorch/pytorch/blob/b5dd6e3e61c59e88ce6e8aee2fa5d08d6c56fd97/torch/nn/modules/activation.py#L831
) Module according to further discussion or to leave the concrete implementation choices to the maintainers.

Overall, it would be great to have this option in the PyTorch code. This would make the code mode cleaner and would allow understanding the different head behaviors without having to locally modify the pytorch functions.

## Alternatives

Current alternatives are:
1) creating multiple MultiHeadAttention modules hardcoded with a single head to retrieve the attention scores of this head (probably less efficient)
2) copying and modifying the [`multi_head_attention_forward`](https://github.com/pytorch/pytorch/blob/16c72a5a6b7085a9651e7bef7e98c80b2d5547e9/torch/nn/functional.py#L4026
) function and [`MultiHeadAttention`](https://github.com/pytorch/pytorch/blob/b5dd6e3e61c59e88ce6e8aee2fa5d08d6c56fd97/torch/nn/modules/activation.py#L831
) Module locally.

cc @albanD @mruberry @zhangguanheng66

### PR DESCRIPTION
**Pull request by jbschlosser**
Fixes #47583

This PR adds a new flag `average_attn_weights` allowing for optionally disabling the averaging of output attention weights across heads. Example usage:
```python
import torch

model = torch.nn.MultiheadAttention(embed_dim=100, num_heads=4)
query = torch.randn(196, 5, 100)
key = torch.randn(196, 5, 100)
value = torch.randn(196, 5, 100)
out, weights = model(query, key, value, average_attn_weights=False)
print(out.shape, weights.shape)
```
```
torch.Size([196, 5, 100]) torch.Size([5, 4, 196, 196])
```

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/modules/activation.h [hunk 1]
```diff
@@ -849,7 +849,8 @@ class TORCH_API MultiheadAttentionImpl
 
   std::tuple<Tensor, Tensor> forward(const Tensor& query, const Tensor& key,
                  const Tensor& value, const Tensor& key_padding_mask = {},
-                 bool need_weights = true, const Tensor& attn_mask = {});
+                 bool need_weights = true, const Tensor& attn_mask = {},
+                 bool average_attn_weights = true);
  protected:
   FORWARD_HAS_DEFAULT_ARGS({3, AnyValue(Tensor())}, {4, AnyValue(true)}, {5, AnyValue(Tensor())})
```
Comment (bhosmer) [inline torch/csrc/api/include/torch/nn/modules/activation.h:855]: I think `average_attn_weights` (arg 6) might need to be added to this

**Review by bhosmer — COMMENTED** [review]
LGTM mod one c++ side change noted inline... only loose end though is that quantized entry point - if we're updating it, do we need to update its callers? 

If you know the answer just LMK and I'll accept after the `FORWARD_HAS_DEFAULT_ARGS` change, looks great otherwise!

File: torch/csrc/api/include/torch/nn/modules/activation.h (hunk 1 above)
Comment (jbschlosser) [reply torch/csrc/api/include/torch/nn/modules/activation.h:855]: nice catch :)

File: torch/nn/quantizable/modules/activation.py [hunk 2]
```diff
@@ -297,15 +298,16 @@ def forward(self,
...
 
     def _forward_impl(self,
                       query: Tensor,
                       key: Tensor,
                       value: Tensor,
                       key_padding_mask: Optional[Tensor] = None,
                       need_weights: bool = True,
-                      attn_mask: Optional[Tensor] = None) -> Tuple[Tensor, Optional[Tensor]]:
+                      attn_mask: Optional[Tensor] = None,
+                      average_attn_weights: bool = True) -> Tuple[Tensor, Optional[Tensor]]:
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply torch/nn/quantizable/modules/activation.py:310]: True, let me ping someone from quantization to help out here. cc @z-a-f @vkuzo are there changes needed beyond this for the new flag?

[the remark this answers is not in the document]
Comment (vkuzo) [reply torch/nn/quantizable/modules/activation.py:310]: sorry for late response, quantizeable MHA is not released to any customers, so as long as tests pass it should be good to go.  Hope that helps!

**Review by bhosmer — APPROVED** [review]
LGTM!

**Pull request closed without merging**